pkg/routine/routine.go

```diff
@@ -0,0 +1,79 @@
+// Copyright © 2019 by PACE Telematics GmbH. All rights reserved.
+// Created at 2019/10/01 by Marius Neugebauer
+
+// Package routine helps in starting background tasks.
+package routine
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/pace/bricks/http/oauth2"
+	"github.com/pace/bricks/maintenance/errors"
+	"github.com/pace/bricks/maintenance/log"
+)
+
+// Forever is so far in the future that it basically never ends.
+var Forever = time.Duration(1<<63 - 1)
+
+// Run runs the given function in a new context with the given timeout. The new
+// context inherits the logger and oauth2 authentication of the passed context.
+// Panics thrown in the function are logged and sent to sentry. The routines
+// context is cancelled if the program receives a shutdown signal (SIGINT,
+// SIGTERM).
+func Run(ctx context.Context, timeout time.Duration, routine func(context.Context)) (cancel context.CancelFunc) {
+	// transfer logger, authentication and error info
+	routineCtx := log.Ctx(ctx).WithContext(context.Background())
+	routineCtx = oauth2.ContextTransfer(ctx, routineCtx)
+	routineCtx = errors.ContextTransfer(ctx, routineCtx)
+	// apply timeout
+	routineCtx, cancel = context.WithTimeout(routineCtx, timeout)
+	// register context to be cancelled when the program is shut down
+	contextsMx.Lock()
+	contexts[routineCtx] = cancel
+	contextsMx.Unlock()
+	// deregister the above if context is done
+	go func() {
```

Comment:
would not start a sep. go routine but move that as a defer func into the func below.

Reply:
If it was a defer in the function below, then it would not be triggered when the context times out, but when the routine finishes.